File: ios/Hotboard/Hotboard/Components/SubmitButtonComponent.swift
import HotwireNative
import UIKit

final class SubmitButtonComponent: BridgeComponent {
    override class var name: String { "submit-button" }

    override func onReceive(message: Message) {
        guard let viewController else { return }
        guard let data: MessageData = message.data() else { return }
        addButton(using: data, to: viewController)
    }

    private var viewController: UIViewController? {
        delegate?.destination as? UIViewController
    }

    private func addButton(
        using data: MessageData,
        to viewController: UIViewController
    ) {
        let action = UIAction { [unowned self] _ in
            self.reply(to: "connect")
        }
        let item = UIBarButtonItem(title: data.title, primaryAction: action)
        viewController.navigationItem.rightBarButtonItem = item
    }
}

private extension SubmitButtonComponent {
    struct MessageData: Decodable {
        let title: String
    }
}

File: android/app/src/main/java/com/masterhotwire/hotboard/components/SubmitButtonComponent.kt
package com.masterhotwire.hotboard.components

import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.ui.platform.ComposeView
import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import dev.hotwire.navigation.fragments.HotwireFragment
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

class SubmitButtonComponent(
    name: String,
    private val bridgeDelegate: BridgeDelegate<HotwireDestination>
) : BridgeComponent<HotwireDestination>(name, bridgeDelegate) {
    private val buttonId = 1
    private val fragment: HotwireFragment
        get() = bridgeDelegate.destination.fragment as HotwireFragment

    override fun onReceive(message: Message) {
        // Handle incoming messages based on the message `event`.
        when (message.event) {
            "connect" -> addButton(message)
            "disconnect" -> removeButton()
            else -> Log.w(
                "ButtonComponent",
                "Unknown event for message: $message"
            )
        }
    }

    private fun addButton(message: Message) {
        // Deserialize the data payload we sent
        // from the submit button controller
        val data = message.data<MessageData>() ?: return

        // Create the view which will render the button
        val button = ComposeView(fragment.requireContext()).apply {
            id = buttonId
            setContent {
                Button(
                    onClick = { replyTo("connect") }
                ) {
                    Text(data.title)
                }
            }
        }
        // Create a layout that will put the button on
        // the right side of the toolbar
        val layoutParams = Toolbar.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        ).apply { gravity = Gravity.END } // END pushes it to the right side

        // Fetch the toolbar
        val toolbar = fragment.toolbarForNavigation()
        // Add the new button
        toolbar?.addView(button, layoutParams)
    }

    private fun removeButton() {
        val toolbar = fragment.toolbarForNavigation()
        val button = toolbar?.findViewById<ComposeView>(buttonId)
        toolbar?.removeView(button)
    }

    // Use kotlinx.serialization annotations to define a serializable
    // data class that represents the incoming message.data json.
    @Serializable
    data class MessageData(
        @SerialName("title") val title: String
    )
}
